domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt

```diff
@@ -35,14 +33,36 @@ class QuestionTrainingController @Inject constructor(
    */
    fun startQuestionTrainingSession(skillIdsList: List<String>): LiveData<AsyncResult<Any?>> {
     return try {
-        val questionsList = retrieveQuestionsForSkillIds(skillIdsList)
-        questionAssessmentProgressController.beginQuestionTrainingSession(questionsList)
+        questionAssessmentProgressController.beginQuestionTrainingSession(
+          retrieveQuestionsForSkillIds(skillIdsList))
         MutableLiveData(AsyncResult.success<Any?>(null))
     } catch (e: Exception) {
       MutableLiveData(AsyncResult.failed(e))
     }
   }
 
+  private fun retrieveQuestionsForSkillIds(skillIdsList: List<String>): DataProvider<List<Question>> {
+    val questionsDataProvider = topicController.retrieveQuestionsForSkillIds(skillIdsList)
+    return dataProviders.transform(TRAINING_QUESTIONS_PROVIDER, questionsDataProvider) {
+      getFilteredQuestionsForTraining(skillIdsList, it, TOTAL_QUESTIONS_PER_TOPIC/skillIdsList.size)
+    }
+  }
+
+  // Attempts to fetch equal number of questions per skill. Removes any duplicates and limits the questions to be
+  // equal to TOTAL_QUESTIONS_PER_TOPIC questions.
+  private fun getFilteredQuestionsForTraining(
+    skillIdsList: List<String>, questionsList: List<Question>, numQuestionsPerSkill: Int): List<Question>{
+    val trainingQuestions = mutableListOf<Question>()
+    for (skillId in skillIdsList) {
+      trainingQuestions.addAll(questionsList.filter {
+        it.linkedSkillIdsList.contains(skillId) &&
+        !trainingQuestions.contains(it)
+      }.take(numQuestionsPerSkill + 1))
+    }
+    return trainingQuestions
+      .take(TOTAL_QUESTIONS_PER_TOPIC)
```

Comment:
Shouldn't there be a random element to this question selection? Otherwise it seems like the current implementation will always yield the same training session.

Reply:
That's how it is done in the web too. Adding randomness increases complexity a significant amount. We also use question difficulty and match that with skill mastery in the web frontend but since we dont have skill mastery in android I havent added that bit. We can add that bit later on to avoid yielding the same training session

Reply:
As discussed offline, we are implementing randomness for Android since we would like to do it in the future for web too